python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1276,16 +1289,28 @@ def _infer_keys_by_input_names(
 
 
 def _infer_keys_by_output_names(
-    node_def: Union["GraphDefinition", OpDefinition], keys_by_output_name: Mapping[str, AssetKey]
+    node_def: Union["GraphDefinition", OpDefinition],
+    keys_by_output_name: Mapping[str, AssetKey],
+    check_specs_by_output_name: Mapping[str, AssetCheckSpec],
 ) -> Mapping[str, AssetKey]:
     output_names = [output_def.name for output_def in node_def.output_defs]
     if keys_by_output_name:
+        overlapping_asset_and_check_outputs = set(keys_by_output_name.keys()) & set(
+            check_specs_by_output_name.keys()
+        )
         check.invariant(
-            set(keys_by_output_name.keys()) == set(output_names),
-            "The set of output names keys specified in the keys_by_output_name argument must "
-            f"equal the set of asset keys outputted by {node_def.name}. \n"
-            f"keys_by_output_name keys: {set(keys_by_output_name.keys())} \n"
-            f"expected keys: {set(output_names)}",
+            not overlapping_asset_and_check_outputs,
+            "The set of output names associated with asset keys and checks overlap:"
+            f" {overlapping_asset_and_check_outputs}",
+        )
+
+        check.invariant(
+            set(keys_by_output_name.keys()) | set(check_specs_by_output_name.keys())
+            == set(output_names),
+            "The union of the set of output names keys specified in the keys_by_output_name and"
+            " check_specs_by_output_name arguments must equal the set of asset keys outputted by"
+            f" {node_def.name}. \nkeys_by_output_name keys:"
+            f" {set(keys_by_output_name.keys())} \nexpected keys: {set(output_names)}",
```

Comment:
I think it is worth putting this complex string construction in a conditional. This could produce a lot of garbage